src/singpass/singpass-helper-ndi.ts

```diff
@@ -0,0 +1,154 @@
+import { AxiosInstance, AxiosProxyConfig } from "axios";
+import * as querystringUtil from "querystring";
+import { createClient } from "../client/axios-client";
+import { JweUtil } from "../util";
+import { SingpassMyInfoError } from "../util/error/SingpassMyinfoError";
+import { logger } from "../util/Logger";
+import { TokenPayload, TokenResponse } from './shared-constants';
+import { Key } from'../util/KeyUtil';
+import { createClientAssertion } from'../util/SigningUtil';
+
+export interface NdiOidcHelperConstructor {
+	oidcConfigUrl: string;
+	clientID: string;
+	redirectUri: string;
+	jweDecryptKey: Key;
+	clientAssertionSignKey: Key;
+	proxyConfig?: AxiosProxyConfig;
+}
+
+interface OidcConfig {
+	issuer: string;
+	authorization_endpoint: string;
+	token_endpoint: string;
+	jwks_uri: string;
+}
+
+export class NdiOidcHelper {
+	private axiosClient: AxiosInstance;
+	private oidcConfigUrl: string;
+	private clientID: string;
+	private redirectUri: string;
+	private jweDecryptKey: Key;
+	private clientAssertionSignKey: Key;
+
+	constructor(props: NdiOidcHelperConstructor) {
+		this.oidcConfigUrl = props.oidcConfigUrl;
+		this.clientID = props.clientID;
+		this.redirectUri = props.redirectUri;
+		this.jweDecryptKey = props.jweDecryptKey;
+		this.clientAssertionSignKey = props.clientAssertionSignKey;
+
+		this.axiosClient = createClient({
+			timeout: 10000,
+			proxy: props.proxyConfig,
+		});
+	}
+
+	public constructAuthorizationUrl = async (
+		state: string,
+		nonce?: string
+	): Promise<string> => {
+		const {data: {authorization_endpoint}} = await this.axiosClient.get<OidcConfig>(this.oidcConfigUrl);
+
+		const queryParams = {
+			state,
+			...(nonce ? { nonce } : {}),
+			redirect_uri: this.redirectUri,
+			scope: "openid",
+			client_id: this.clientID,
+			response_type: "code",
+
+		};
+		const queryString = querystringUtil.stringify(queryParams);
+		return `${authorization_endpoint}?${queryString}`;
+	}
+
+	/**
+	 * Get tokens from Singpass endpoint. Note: This will fail if not on an IP whitelisted by SP.
+	 * Use getIdTokenPayload on returned Token Response to get the token payload
+	 */
+	public getTokens = async (authCode: string): Promise<TokenResponse> => {
+		const { data: { token_endpoint, issuer } } = await this.axiosClient.get<OidcConfig>(this.oidcConfigUrl);
+
+		const params = {
+			grant_type: "authorization_code",
+			code: authCode,
+			client_id: this.clientID,
+			redirect_uri: this.redirectUri,
+			client_assertion_type: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
+			client_assertion: await createClientAssertion({
+				issuer: this.clientID,
+				subject: this.clientID,
+				audience: issuer,
+				key: this.clientAssertionSignKey,
+			})
+		};
+		const body = querystringUtil.stringify(params);
+
+		const config = {
+			headers: {
+				"content-type": "application/x-www-form-urlencoded; charset=ISO-8859-1"
```

Comment:
question: is `; charset=ISO-8859-1` necessary? or can we

```suggestion
				"content-type": "application/x-www-form-urlencoded"
```